httpd/connlist.c

```diff
@@ -0,0 +1,134 @@
+#include "connlist.h"
+
+conn_list_t conn_list = { NULL, NULL, 0, 0, PTHREAD_MUTEX_INITIALIZER };
+
+static h2o_stream_conn_t **conn_list_get_null_element_unsafe(conn_list_t *list)
+{
+    struct conn_list_leaf *leaf = list->head;
+    while (leaf != NULL) {
+        for (int i = 0; i < CONN_LIST_MEMPOOL_SIZE; i++) {
+            if (leaf->conn[i] == NULL)
+                return &leaf->conn[i];
+        }
+        leaf = leaf->next;
+    }
+    return NULL;
+}
+
+void conn_list_insert(conn_list_t *list, h2o_stream_conn_t *conn)
+{
+    pthread_mutex_lock(&list->lock);
+
+    // in case the allocated capacity is not used up
+    // we can reuse the null element
+    if (list->capacity != list->size) {
+        h2o_stream_conn_t **null_element = conn_list_get_null_element_unsafe(list);
+        if (unlikely(null_element == NULL)) {
+            error_report("conn_list_insert: capacity != size but no null element found");
+            return;
+        }
+        *null_element = conn;
+        list->size++;
+        pthread_mutex_unlock(&list->lock);
+        return;
+    }
+
+    // if not, we need to allocate a new leaf
+    struct conn_list_leaf *old_tail = list->tail;
+    list->tail = callocz(1, sizeof(struct conn_list_leaf));
```

Comment:
Considering that we are going to allocate different structures here with the same size, we have a perfect scenario to use our [aral](https://github.com/netdata/netdata/blob/master/libnetdata/aral/aral.c) lib, take a look when you have a time.

Reply:
I resolved all other comments from you. This one is worth thinking about but if I decide to convert to ARAL then it will be separate PR